File: adminsite/addplayer.php
/*
 *  Copyright 2006-2008 Jerome PASQUIER
 * 
 *  This file is part of RainbruRPG.
 *
 *  RainbruRPG is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  RainbruRPG is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with RainbruRPG; if not, write to the Free Software
 *  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
 *  02110-1301  USA
 *
 */

<?php
/** \file addbonusfilemodifier.php
  * Defines a page used to add a player
  * 
  * Modifications :
  * - 26 sep 2008 : Single documentation added
  *
  */

  /// Get the access rights
$acc=include 'access.php';
if ($acc){
?>

<?php
 include "xmlplayerinterface.php";

/** Adds a player
  *
  */
function formaddplayer(){
  $err=false;

  $name = trim($_POST['name']);
  $pwd = trim($_POST['pwd']);
  $mail = trim($_POST['mail']);
  $timestamp = trim($_POST['timestamp']);

  $xmlplayer= new XmlPlayerInterface();

  // A field has been filled
  if(!empty($name)|!empty($pwd)|!empty($mail)|!empty($timestamp)){
    $errPhrase='<div class="error">';

    if (empty($name)){
      $err=true;
      $errPhrase=$errPhrase.'The name can not be empty.<br>';
    }
    if (empty($pwd)){
      $err=true;
      $errPhrase=$errPhrase."The password can not be empty.<br>";
    }
    if (empty($timestamp)){
      $err=true;
      $errPhrase=$errPhrase."The timestamp can not be empty.<br>";
    }
    if (empty($mail)){
      $err=true;
      $errPhrase=$errPhrase."The mail adress field can not be empty.<br>";
    }
    else{
      // Tests the mail adress
      if (strstr($mail, '@')==FALSE){
	$err=true;
	$errPhrase=$errPhrase."The mail adress is invalid (@ missing).<br>";
      }
      if (strstr($mail, '.')==FALSE){
	$err=true;
	$errPhrase=$errPhrase."The mail adress is invalid (. missing).<br>";
      }
    }
    if ($xmlplayer->isPlayerExists($name)){
      $err=true;
      $errPhrase=$errPhrase."This player's name is already in use.<br>";
    }
    if ($xmlplayer->isMailExists($mail)){
      $err=true;
      $errPhrase=$errPhrase."This mail adress is already in use.<br>";
    }
    echo("</div>");
    // Last test before adding player
    if ($err){
      echo $errPhrase=$errPhrase;
      echo('<font text="red"><p>The player can not be added.</p></font>');
     }
    else{
      echo 'The player is added';

      $confirmId=md5(uniqid(mt_rand()));
      $xmlplayer->addPlayer($name, $mail, $pwd, $timestamp, $confirmId);


      // Send a confirmation mail
      $confirmLink ="http://rainbru.free.fr/rpg/admin/mailconfirm.php?name=";
      $confirmLink.=$name."&id=".$confirmId;
      sendConfirmationMail($name, $mail, $confirmLink);

    }
  }
}

 function sendConfirmationMail($name, $mail, $confirmLink){
   //-----------------------------------------------
   //DECLARE LES VARIABLES
   //-----------------------------------------------
   $sujet='Account verification';
   $destinataire=$mail;
   $email_expediteur='rainbru@free.fr';
   $email_reply='rainbru@free.fr';
     
   $message ='
     <html>
       <head>
         <title>RainbruRPG account registration</title>
       </head>
       <body>
         Hi '.$name.','.'<br>
         <br>
         You have requested an account creation for RainbruRPG, thanks. 
         Please go to the following web page to confirm that this 
         mail address is valid.<br>
         <a href="'.$confirmLink.'">'.$confirmLink.'</a><br>
         <br>
         If you haven'."'".'t requested anything, please contact me :<br>
         <a href="mailto:'.$email_reply.'">'.$email_reply.'</a><br>
       </body>
     </html>';

   $headers ='From: "RainbruRPG admin"<'.$email_expediteur.'>'."\n";
   $headers .='Reply-To: '.$email_reply."\n";
   $headers .='Content-Type: text/html; charset="iso-8859-1"'."\n";
   $headers .='Content-Transfer-Encoding: 8bit';
   

   if(mail($destinataire,$sujet,$message,$headers)){
     echo '<br>The mail was correctly sent';
   }
   else{
     echo '<br>An error occured. The message was not sent.';
   } 

}


?>

<HTML>
  <head>
    <title>phpIdent</title>
    <link href="content.css" rel="stylesheet" type="text/css">
  </head>
  <body>
    <h1>Create account</h1>
    <p>Here, you can add a player to play RainbruRPG.</p>
    <form action="addplayer.php" method="post">
<div align='center'>
<table width='400px' class="noborder">
<tr>
      <td>Player name </td>
      <td><input type="text" name="name" /></td>
</tr>
<tr>
      <td>Password hashsum</td> 
      <td><input type="text" name="pwd" /></td>
</tr>
<tr>
      <td>Creation timestamp</td> 
      <td><input type="text" name="timestamp" /></td>
</tr>
<tr>
      <td>EMail</td> 
      <td><input type="text" name="mail" /></td>
</tr>
</table>
      <p><input type="submit" value="OK"></p>
</div>


    </form>
<?php
   printf('<p>Now, the timestamp is : %d</p>', time()); 
formaddPlayer();
?>
  </body>
</HTML>


<?php } ?>

File: rainbrurpg/tools/oldXident/addplayer.cpp
/*
 *  Copyright 2006 Jerome PASQUIER
 *
 *  This file is part of RainbruRPG.
 *
 *  RainbruRPG is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  RainbruRPG is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with RainbruRPG; if not, write to the Free Software
 *  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
 *  02110-1301  USA
 *
 */

#include "addplayer.h"

#include "playerlist.h"


/** The default constructor
  *
  * The parameter must be no-NULL because it is used to refresh the
  * player list.
  *
  * \param pl The not NULL instance of PlayerList calling this dialog.
  */
RainbruRPG::Gui::AddPlayer::AddPlayer(PlayerList* pl)
                          :Fl_Window( 400, 300, "Create a player") {

  playerList=pl;

  srand( (unsigned) time( NULL ));
  cars="abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789";

    name =new Fl_Input( 100, 20, 100, 30, "Name");
    password =new Fl_Output( 100, 60, 100, 30, "Password");
    mail =new Fl_Input( 100, 100, 250, 30, "Mail adress");

    Fl_Button* cancelBtn = new Fl_Button( 10, 260, 100, 30, "Cancel" );
    Fl_Button* rollBtn = new Fl_Button( 200-55, 260, 100, 30, "Roll" );
    Fl_Button* okBtn = new Fl_Button( 400-110, 260, 100, 30, "OK" );

  end();

  cancelBtn->callback(cbCancel, this);
  rollBtn->callback(cbRoll, this);
  okBtn->callback(cbOk, this);
  this->set_modal();

  this->show();
  roll();
}

/** The default destructor 
  *
  */
RainbruRPG::Gui::AddPlayer::~AddPlayer(){
  name       =NULL;
  password   =NULL;
  playerList =NULL;
}

/** The static callback for the note's text multiline input
  *
  * It simply calls the i_cbRoll() method.
  *
  * \param w A parameter used internally by fltk
  * \param f A parameter used internally by fltk
  */
void RainbruRPG::Gui::AddPlayer::cbRoll(Fl_Widget *w, void *f){
  ( (AddPlayer*)f )->i_cbRoll();
}

/** The callback of the roll button
  *
  * It simply calls the roll() method.
  */
void RainbruRPG::Gui::AddPlayer::i_cbRoll(){
  roll();
}

/** The static callback for the note's text multiline input
  *
  * It simply calls the i_cbCancel() method.
  *
  * \param w A parameter used internally by fltk
  * \param f A parameter used internally by fltk
  */
void RainbruRPG::Gui::AddPlayer::cbCancel(Fl_Widget *w, void *f){
  ( (AddPlayer*)f )->i_cbCancel();
}

/** The callback of the cancel button
  *
  * It calls the razDialog() and the hide() methods.
  */
void RainbruRPG::Gui::AddPlayer::i_cbCancel(){
  LOGI("Hidding AddPlayer dialog");
  this->razDialog();
  this->hide();
}

/** The static callback for the note's text multiline input
  *
  * It simply calls the i_cbOk() method.
  *
  * \param w A parameter used internally by fltk
  * \param f A parameter used internally by fltk
  */
void RainbruRPG::Gui::AddPlayer::cbOk(Fl_Widget *w, void *f){
  ( (AddPlayer*)f )->i_cbOk();
}

/** The callback of the OK button
  *
  * It tests if the name has at least 4 character and create the
  * player with xmlPlayerList. If the name already exists, it can not
  * create it.
  */
void RainbruRPG::Gui::AddPlayer::i_cbOk(){
  CurlPlayerAdd tpa;
  tpa.setPostedData("name", name->value());
  tpa.setPostedPwd("pwd", password->value());
  tpa.setPostedData("mail", mail->value());

  bool success=tpa.perform();

  if (!success)
    ShowCurlError::showError(&tpa);

  playerList->refresh();
  this->hide();
}

/** Randomly generate a new password
  *
  * I use the cars string, initialized in the constructor with letters
  * and numbers. I randomly take a character in the cars string and
  * concatenate them to obtain a string of 8 random charachers. It
  * uses the random(int) function to get a random integer.
  *
  */
void RainbruRPG::Gui::AddPlayer::roll(){
  // We take a random number between 0 and cars.lenght()-1

  string mdp;

  for (int j=0; j<8; j++){
    int i=random(cars.length());
    char c = cars[i];
    mdp+=c;
  }
  password->value( mdp.c_str() );
}

/** Returns a random integer between 0 and \c max
  *
  * \param max the upper limit of the random number.
  * \return An integer between 0 and the integer parameter.
  */
int RainbruRPG::Gui::AddPlayer::random(int max){
  int a=(int) ((float)max * (float)rand() / (float)RAND_MAX);
  return a;
}

/** Make the dialog empty
  * 
  */
void RainbruRPG::Gui::AddPlayer::razDialog(){
  name->value("");
  mail->value("");
}
